src/runtime-rs/crates/resource/src/network/utils/link/mod.rs

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+mod create;
+pub use create::{create_link, LinkType};
+mod driver_info;
+pub use driver_info::{get_driver_info, DriverInfo};
+mod macros;
+mod manager;
+pub use manager::get_link_from_message;
+
+use std::os::unix::io::RawFd;
+
+use netlink_packet_route::link::nlas::State;
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum Namespace {
+    NetNsPid(u32),
+    #[allow(dead_code)]
+    NetNsFd(RawFd),
+}
+impl Default for Namespace {
+    fn default() -> Self {
+        Self::NetNsPid(0)
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum LinkStatistics {
+    #[allow(dead_code)]
+    Stats(LinkStatistics32),
+    Stats64(LinkStatistics64),
+}
+impl Default for LinkStatistics {
+    fn default() -> Self {
+        Self::Stats64(LinkStatistics64::default())
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Default)]
+pub struct LinkStatistics32 {
+    pub rx_packets: u32,
+    pub tx_packets: u32,
+    pub rx_bytes: u32,
+    pub tx_bytes: u32,
+    pub rx_errors: u32,
+    pub tx_errors: u32,
+    pub rx_dropped: u32,
+    pub tx_dropped: u32,
+    pub multicast: u32,
+    pub collisions: u32,
+    pub rx_length_errors: u32,
+    pub rx_over_errors: u32,
+    pub rx_crc_errors: u32,
+    pub rx_frame_errors: u32,
+    pub rx_fifo_errors: u32,
+    pub rx_missed_errors: u32,
+    pub tx_aborted_errors: u32,
+    pub tx_carrier_errors: u32,
+    pub tx_fifo_errors: u32,
+    pub tx_heartbeat_errors: u32,
+    pub tx_window_errors: u32,
+    pub rx_compressed: u32,
+    pub tx_compressed: u32,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Default)]
+pub struct LinkStatistics64 {
```

Comment:
- You could use a macro to define these types to avoid the duplication.
- Are these well known structures that you can add a kernel URL for? If so, are they defined in newer versions of `libc` / `nix`?

Reply:
Thanks for your review, raise issue #4154 to fix it.